core/src/main/scala/org/apache/spark/scheduler/cluster/CoarseGrainedSchedulerBackend.scala

```diff
@@ -413,25 +413,38 @@ class CoarseGrainedSchedulerBackend(scheduler: TaskSchedulerImpl, val rpcEnv: Rp
    *
    * @param executorIds identifiers of executors to kill
    * @param replace whether to replace the killed executors with new ones
+   * @param force whether to kill busy executors (who are running tasks)
    * @return whether the kill request is acknowledged.
    */
-  final def killExecutors(executorIds: Seq[String], replace: Boolean): Boolean = synchronized {
+  final def killExecutors(
+      executorIds: Seq[String],
+      replace: Boolean,
+      force: Boolean): Boolean = synchronized {
     logInfo(s"Requesting to kill executor(s) ${executorIds.mkString(", ")}")
     val (knownExecutors, unknownExecutors) = executorIds.partition(executorDataMap.contains)
     unknownExecutors.foreach { id =>
       logWarning(s"Executor to kill $id does not exist!")
     }
 
+    val idleExecutors = {
+      if (force) {
+        knownExecutors
+      } else {
+        knownExecutors.filter(executor =>
+          !scheduler.taskIdToExecutorId.exists(_._2 == executor))
```

Comment:
This would be somewhat expensive if you have 50k tasks, and we iterate through this once per executor we want to kill. Maybe in `TaskSchedulerImpl` we should keep track of a set of executors that currently have tasks scheduled on them instead, so this would take O(num executors) instead of O(num executors * num active tasks).